source/extensions/filters/http/jwt_authn/filter_config.h

```diff
@@ -68,8 +68,18 @@ class FilterConfig : public Logger::Loggable<Logger::Id::jwt>, public AuthFactor
         tls_(context.threadLocal().allocateSlot()), cm_(context.clusterManager()),
         time_source_(context.dispatcher().timeSource()), api_(context.api()) {
     ENVOY_LOG(info, "Loaded JwtAuthConfig: {}", proto_config_.DebugString());
-    tls_->set([this](Event::Dispatcher&) -> ThreadLocal::ThreadLocalObjectSharedPtr {
-      return std::make_shared<ThreadLocalCache>(proto_config_, time_source_, api_);
+
+    // note: `this` and `context` has a a lifetime of the listener.
+    // that may be shorter of the tls callback if the listener is torn shortly after it is created.
+    // context.dispatcher().timeSource() and context.api() are from the server context who's
+    // lifetime is longer. so we use them explicitly.
+    // we copy over the proto as we can't guarantee its lifetime otherwise
+    auto& timeSource = context.dispatcher().timeSource();
+    auto& api = context.api();
+    auto proto_config_copy = proto_config_;
+    tls_->set([proto_config_copy, &timeSource,
```

Comment:
Hmm,  my understanding is tls_ own these calls. if  tls_ is gone, these calls should be removed, should not be called any more.   FilterConfig owns tls_  so when FilterConfig is gone,  tls_ should be gone, and its registered calls should not be called any more.  

is this assumption correct?

Reply:
Just read the thread local impl code.  I guess my assumption is not correct.  the callback is stored in each thread dispatcher.   it could still be called even the slot is gone.

Reply:
tls_ sends the calls to the worker threads. so in that sense there can always be a race.
i.e. even if tls_ unregisters the some of the calls during destruction, there is no guarantee that there is no call in progress running in a different thread..

Reply:
Can you do a quick audit of other extensions? It wouldn't surprise me if we had this same bug elsewhere. I would love to fix this more systemically but I'm not sure how off the top of my head.

Reply:
@mattklein123 i think i found another problematic usage:
https://github.com/envoyproxy/envoy/blob/46eb08454672710fb77256bca3581927c146da8c/source/extensions/filters/network/redis_proxy/conn_pool_impl.cc#L51

Reply:
OK thanks for looking. Let me spend a bit of time later today trying to think about whether there is some holistic way to fix this.

Reply:
one thought that comes to mind is to use enable_shared_from_this for these cases, so that the instances can pass shared pointer to themselves.

Reply:
I think this idea.  Could you use it for jwt_authn too?

Reply:
@mattklein123 @yuval-k So is the plan here in this PR is to fix just  `jwt_authn` filter and later look for holistic implementation? Or fix it in a separate PR following the same approach for Redis? The reason I am asking this is, we have seen Redis listener bind issues with port conflicts caused Envoy to crash some times. I was thinking some thing like this might be the issue, but could n't dig deep at that time. But looking at it now, it looks like this is the reason why it is crashing. So would like this to be fixed in  Redis as well.

@yuval-k Thanks a lot for digging in to this .